server/resp/src/main/java/org/infinispan/server/resp/tx/RespTransactionHandler.java

```diff
@@ -67,13 +71,20 @@ protected CompletionStage<RespRequestHandler> actualHandleRequest(ChannelHandler
       // This method already writes any error message to the socket.
       if (!isCommandValid(command, arguments)) return myStage();
 
-      queued.add(new TransactionCommand(command, List.copyOf(arguments)));
+      try {
+         queued.add(new TransactionCommand(command, List.copyOf(arguments)));
+      } catch (Throwable t) {
+         errorInTransactionContext();
+         return command.handleException(this, t);
+      }
+
       return stageToReturn(myStage(), ctx, Consumers.QUEUED_BICONSUMER);
    }
 
    @Override
    public void handleChannelDisconnect(ChannelHandlerContext ctx) {
-      dropTransaction();
+      // Possibly blocking operation. De-register transaction watchers.
```

Comment:
This sounds like an issue.. do we really need to wait for the listeners to be removed here?